User profile - release notes and more migration instructions (#27003)
keycloak/keycloak: docs/documentation/release_notes/topics/24_0_0.adoc

= Supported user profile and progressive profiling

The user profile preview feature is promoted to be fully supported and user profile is enabled by default!

In the past months, Keycloak team spent huge amount of effort in polishing user
profile and making it fully supported now. In this release, we continued the effort started in the last release. Lots of improvements, fixes and
polishing was done based on the thorought testing and feedback from our awesome community.

Few highlights of the user profile feature:

> 

Reviewer: ```suggestion
* Fine-grained control over the attributes that users and administrators can manage so that you can prevent unexpected attributes, and values, from being set. 
```

* Ability to specify what user attributes are managed and should be displayed on the forms to regular users or administrators

* Dynamic forms - Previously, the forms where users created or updated their profiles, contain four basic attributes like username, email, first name and last name. Any additional
attributes (or removing some of the default attributes) required to create custom theme. Now custom themes may not be needed anymore as users will see exactly the requested attributes based on the requirement
of the particular deployment

* Validations - Ability to specify validators for the user attributes. Lot of built-in validators allowing to specify maximum/minimum length, specified regex, fact
that particular attribute must be URL or number and lots of other options

> * Annotations - Ability to specify that particular attribute should be rendered for instance as text area or HTML select with specified options or calendar and lots of other options

Reviewer: ```suggestion
* Annotations - Ability to specify that particular attribute should be rendered for instance as text area or HTML select with specified options or calendar and lots of other options. You can also bind JavaScript code to a specific field to change how an attribute is rendered and customize its behavior.
```

* Progressive profiling - Allows to specify that some fields are required or available on the forms just for particular values of `scope` parameter. This effectively allow progressive
profiling. No need anymore to ask user for twenty attributes during registration since you can instead ask user to fill attributes incrementally according to the requirements of individual client
applications, which user want to use.

* Migration from previous versions - User profile is now always enabled, but things should still work as before for those, who did not used user profile until now. You can
benefit from user profile capabilities, but you are not strictly forced to use them. For migration instructions consult the link:{upgradingguide_link}[{upgradingguide_name}].

> 

Reviewer: This first release of the user profile as a supported feature is just the starting point and the baseline for delivering many more capabilities around identity management.

We would like to huge Thanks to awesome Keycloak community as lots of ideas, requirements and contributions came from the community! For more details about user profile capabilities, check
link:{adminguide_link}#user-profile[{adminguide_name}].

> 

Reviewer: Special thanks to:

* https://github.com/velias
* https://github.com/alechenninger
* https://github.com/thomasdarimont
* https://github.com/bs-matil
* https://github.com/sschu
* https://github.com/antikalk
* https://github.com/patrickjennings

= Supported user profile and progressive profiling

The user profile preview feature is promoted to be fully supported and user profile is enabled by default!

> In the past months, Keycloak team spent huge amount of effort in polishing user
> profile and making it fully supported now. In this release, we continued the effort started in the last release. Lots of improvements, fixes and
> polishing was done based on the thorought testing and feedback from our awesome community.

Reviewer 2: ```suggestion
In the the past months, Keycloak team spent a huge amount of effort in polishing the user
profile feature to make it fully supported. In this release, we continued the effort that was started in the last release. Lots of improvements, fixes and
polishing was done based on the thorough testing and feedback from our awesome community.
```

Few highlights of the user profile feature:
* Fine-grained control over the attributes that users and administrators can manage so that you can prevent unexpected attributes, and values, from being set. 
* Ability to specify what user attributes are managed and should be displayed on the forms to regular users or administrators

* Dynamic forms - Previously, the forms where users created or updated their profiles, contain four basic attributes like username, email, first name and last name. Any additional

> attributes (or removing some of the default attributes) required to create custom theme. Now custom themes may not be needed anymore as users will see exactly the requested attributes based on the requirement

Reviewer 2: ```suggestion
attributes (or removing some of the default attributes) required to create custom theme. Now custom themes may no longer be needed because users see exactly the requested attributes based on the requirement
```

of the particular deployment

> * Validations - Ability to specify validators for the user attributes. Lot of built-in validators allowing to specify maximum/minimum length, specified regex, fact
> that particular attribute must be URL or number and lots of other options

Reviewer 2: ```suggestion
* Validations - Ability to specify validators for the user attributes including built-in validators that you can use to specify a maximum or minimum length, a specific regex, or limiting an 
particular attribute to be a URL or number.
```

* Annotations - Ability to specify that particular attribute should be rendered for instance as text area or HTML select with specified options or calendar and lots of other options. You can also bind JavaScript code to a specific field to change how an attribute is rendered and customize its behavior.

> * Progressive profiling - Allows to specify that some fields are required or available on the forms just for particular values of `scope` parameter. This effectively allow progressive

Reviewer 2: ```suggestion
* Progressive profiling - Ability to specify that some fields are required or available on the forms just for particular values of `scope` parameter. This effectively allow progressive
```

> profiling. No need anymore to ask user for twenty attributes during registration since you can instead ask user to fill attributes incrementally according to the requirements of individual client
> applications, which user want to use.

Reviewer 2: ```suggestion
profiling. You no longer need to ask user for twenty attributes during registration; you can instead ask the user to fill in attributes incrementally according to the requirements of the individual client
applications that is used by the user.
```

> * Migration from previous versions - User profile is now always enabled, but things should still work as before for those, who did not used user profile until now. You can
> benefit from user profile capabilities, but you are not strictly forced to use them. For migration instructions consult the link:{upgradingguide_link}[{upgradingguide_name}].

Reviewer 2: ```suggestion
* Migration from previous versions - User profile is now always enabled, but it operates as before for those who did not use this feature till now. You can
benefit from the user profile capabilities, but you are not strictly forced to use them. For migration instructions consult the link:{upgradingguide_link}[{upgradingguide_name}].
```

This first release of the user profile as a supported feature is just the starting point and the baseline for delivering many more capabilities around identity management.

We would like to huge Thanks to awesome Keycloak community as lots of ideas, requirements and contributions came from the community! Special thanks to:

* https://github.com/velias[Vlastimil Eliáš]
* https://github.com/alechenninger[Alec Henninger]
* https://github.com/thomasdarimont[Thomas Darimont]
* https://github.com/bs-matil[Markus Till]
* https://github.com/sschu[Sebastian Schuster]
* https://github.com/antikalk[Oliver]
* https://github.com/patrickjennings[Patrick Jennings]
* https://github.com/adrhine[Andrew]

For more details about user profile capabilities, check link:{adminguide_link}#user-profile[{adminguide_name}].

== Breaking changes to the User Profile SPI

> In this release, there are changes to the User Profile SPI that might impact existing implementations based on this SPI. For more details, check the

Reviewer 2: ```suggestion
In this release, changes to the User Profile SPI might impact existing implementations based on this SPI. For more details, check the
```

link:{upgradingguide_link}[{upgradingguide_name}].

== Changes to Freemarker templates to allow rendering pages based on the user profile configuration set to a realm

In this release, the following templates were updated to make it possible to dynamically render attributes based
on the user profile configuration set to a realm:

* `login-update-profile.ftl`
* `register.ftl`
* `update-email.ftl`

For more details, see link:{upgradingguide_link}[{upgradingguide_name}].

> == The update profile page when logging in for the first time through a broker now have its own Freemarker templates

Reviewer 2: ```suggestion
== The update profile page when logging in for the first time through a broker now has its own Freemarker templates
```

> In this release, the server will render the update profile page when the user is authenticating through a broker for the

Reviewer 2: ```suggestion
In this release, the server renders the update profile page when the user is authenticating through a broker for the
```
